### PR TITLE
Application Refactoring

### PR DESCRIPTION
**Pull request by sidz**
This PR brings a lot of changes to PHPMND.
- [x] Generators
- [x] Objects (DetectionResult) instead of plain array
- [x] Container
- [x] More tests
- [x] No BC breaks  

Reducing memory usage and speed up scanning a bit. Real example on my current project:
current master:
```
Total of Magic Numbers: 17
Time: 00:06.161, Memory: 32.00 MB
```
This PR:
```
Total of Magic Numbers: 17
Time: 00:05.373, Memory: 28.00 MB
```

### CODE REVIEW

**Comment by exussum12** [comment]
Still going though these, I will mark as hacktober accepted in case you are participating, changes look good though

**Comment by sidz** [comment]
@exussum12 sorry for disturbing. how is review going?

**Comment by exussum12** [comment]
Apologies I've been away. I can take a proper look tomorrow

File: src/Printer/Console.php [hunk 1]
```diff
@@ -4,52 +4,83 @@
...
-            $total += count($entries);
-            foreach ($entries as $entry) {
+        foreach ($this->groupDetectionResultPerFile($list) as $detectionResults) {
+            foreach ($detectionResults as $detection) {
                 $output->writeln(sprintf(
-                    '%s:%d  Magic number: %s',
-                    $fileReport->getFile()->getRelativePathname(),
-                    $entry['line'],
-                    $entry['value']
+                    '%s:%d. Magic number: %s',
```
Comment (exussum12) [inline src/Printer/Console.php:35]: This is a breaking change, its probably better but we should wait till 3 really, Its part of the output

File: src/Printer/Console.php [hunk 2]
```diff
@@ -4,52 +4,83 @@
...
+     * @var Highlighter
+     */
+    private $highlighter;
+
+    public function __construct(Highlighter $highlighter)
+    {
+        $this->highlighter = $highlighter;
+    }
+
+    public function printData(OutputInterface $output, HintList $hintList, array $list): void
```
Comment (exussum12) [inline src/Printer/Console.php:26]: what is list here ? Can we name it better?

File: src/Printer/Console.php [hunk 3]
```diff
@@ -4,52 +4,83 @@
...
+
+        $output->writeln('<info>Total of Magic Numbers: ' . count($list) . '</info>');
+    }
+
+    /**
+     * @param array<int, DetectionResult> $list
+     *
+     * @return array<int, DetectionResult[]>
+     */
+    private function groupDetectionResultPerFile(array $list): array
```
Comment (exussum12) [inline src/Printer/Console.php:76]: same here, $list could be $foundMagicNumbers or similar

**Comment by exussum12** [comment]
This is a mega refactor! Im going though the list now to try and verify its not caused any issues. Nice work some really nice changes in there

File: src/Printer/Console.php (hunk 1 above)
Comment (sidz) [reply src/Printer/Console.php:35]: I'd say that it was more like a typo here.

File: src/Printer/Console.php (hunk 2 above)
Comment (sidz) [reply src/Printer/Console.php]: done

File: src/Printer/Console.php (hunk 3 above)
Comment (sidz) [reply src/Printer/Console.php]: done

**Review by exussum12 — APPROVED** [review]

**Comment by exussum12** [comment]
Approved. Nice work

**Comment by sidz** [comment]
Thanks @exussum12

**Pull request merged**